File: src/game/cipherCursor.ts
import { SymbolMapping } from '../types';

export function dismissMobileKeyboard(input: HTMLInputElement | null) {
  input?.blur();
  const active = document.activeElement;
  if (active instanceof HTMLElement) active.blur();
}

export function selectedGlyphTile(symbolId: string, cellId?: string | null) {
  if (cellId) {
    const cell = document.querySelector(`[data-cipher-cell="${CSS.escape(cellId)}"]`);
    if (cell instanceof HTMLElement && cell.getAttribute('data-cipher-symbol') === symbolId) {
      return cell;
    }
  }
  return document.querySelector(`[data-cipher-symbol="${CSS.escape(symbolId)}"]`);
}

export function placeCipherInput(symbolId: string, input: HTMLInputElement, cellId?: string | null) {
  const tile = selectedGlyphTile(symbolId, cellId);
  if (!(tile instanceof HTMLElement)) return null;
  const rect = tile.getBoundingClientRect();
  input.style.left = `${rect.left}px`;
  input.style.top = `${rect.top}px`;
  input.style.width = `${Math.max(16, rect.width)}px`;
  input.style.height = `${Math.max(16, rect.height)}px`;
  return tile;
}

export function letterCells(
  words: { id: string; symbols: { symbolId: string; isPunctuation?: boolean }[] }[]
) {
  const cells: { cellId: string; symbolId: string }[] = [];
  words.forEach((word) => {
    word.symbols.forEach((item, charIdx) => {
      if (!item.isPunctuation) cells.push({ cellId: `${word.id}_${charIdx}`, symbolId: item.symbolId });
    });
  });
  return cells;
}

export function cellCursor(
  cells: { cellId: string; symbolId: string }[],
  selectedSymbolId: string | null,
  selectedCellId?: string | null
) {
  if (selectedCellId) {
    const byCell = cells.findIndex((cell) => cell.cellId === selectedCellId);
    if (byCell >= 0) return byCell;
  }
  if (!selectedSymbolId) return 0;
  const bySymbol = cells.findIndex((cell) => cell.symbolId === selectedSymbolId);
  return bySymbol < 0 ? 0 : bySymbol;
}

export function nextOpenCell(
  cells: { cellId: string; symbolId: string }[],
  selectedSymbolId: string | null,
  selectedCellId: string | null,
  mappings: SymbolMapping
) {
  if (!cells.length) return null;
  const start = cellCursor(cells, selectedSymbolId, selectedCellId);
  for (let step = 1; step <= cells.length; step += 1) {
    const cell = cells[(start + step) % cells.length];
    if (!mappings[cell.symbolId]) return cell;
  }
  return null;
}

export function previousCell(
  cells: { cellId: string; symbolId: string }[],
  selectedSymbolId: string | null,
  selectedCellId?: string | null
) {
  const start = cellCursor(cells, selectedSymbolId, selectedCellId);
  if (start <= 0) return null;
  return cells[start - 1];
}

export function webTypeFeel(kind: 'tap' | 'key') {
  try {
    navigator.vibrate?.(kind === 'tap' ? 12 : 8);
  } catch {
    return;
  }
}

File: android/core/cipher/src/main/kotlin/com/chroniclecryptogram/cipher/CipherCursor.kt
package com.chroniclecryptogram.cipher

import com.chroniclecryptogram.cipher.model.CryptogramWord

/** One selectable position on the board. Ids are `<wordId>_<indexWithinWord>`. */
data class BoardCell(
    val cellId: String,
    val symbolId: String,
)

/**
 * Cursor arithmetic, ported from the pure half of `src/game/cipherCursor.ts`
 * (lines 32-79). Pinned by the cursor fixture.
 *
 * The DOM half of that file -- `placeCipherInput`, `selectedGlyphTile`,
 * `dismissMobileKeyboard` -- has no counterpart here and never will: it existed
 * to park a hidden input over the selected tile, which the Compose board solves
 * by reporting real placement rectangles out of its own layout pass.
 */
object CipherCursor {

    /** Every non-punctuation position, in reading order. */
    fun letterCells(words: List<CryptogramWord>): List<BoardCell> {
        val cells = ArrayList<BoardCell>()
        for (word in words) {
            word.symbols.forEachIndexed { index, cell ->
                if (!cell.isPunctuation) {
                    cells += BoardCell(cellId = "${word.id}_$index", symbolId = cell.symbolId)
                }
            }
        }
        return cells
    }

    /**
     * Index of the current position. A known cell id wins over a symbol id --
     * the same glyph appears many times, and the player selected one of them.
     * Anything unrecognised falls back to the start of the board.
     */
    fun cellCursor(
        cells: List<BoardCell>,
        selectedSymbolId: String?,
        selectedCellId: String? = null,
    ): Int {
        if (selectedCellId != null) {
            val byCell = cells.indexOfFirst { it.cellId == selectedCellId }
            if (byCell >= 0) return byCell
        }
        if (selectedSymbolId == null) return 0
        val bySymbol = cells.indexOfFirst { it.symbolId == selectedSymbolId }
        return if (bySymbol < 0) 0 else bySymbol
    }

    /**
     * The next cell whose symbol has no guess yet, wrapping past the end.
     *
     * Returns null when every symbol is mapped, which is what stops the search
     * looping forever on a full board.
     */
    fun nextOpenCell(
        cells: List<BoardCell>,
        selectedSymbolId: String?,
        selectedCellId: String?,
        mappings: Map<String, String>,
    ): BoardCell? {
        if (cells.isEmpty()) return null
        val start = cellCursor(cells, selectedSymbolId, selectedCellId)
        for (step in 1..cells.size) {
            val cell = cells[(start + step) % cells.size]
            if (mappings[cell.symbolId].isNullOrEmpty()) return cell
        }
        return null
    }

    /** The previous cell, or null at the start of the board. Does not wrap. */
    fun previousCell(
        cells: List<BoardCell>,
        selectedSymbolId: String?,
        selectedCellId: String? = null,
    ): BoardCell? {
        val start = cellCursor(cells, selectedSymbolId, selectedCellId)
        if (start <= 0) return null
        return cells[start - 1]
    }
}
